crates/starknet/src/state/mod.rs

```diff
@@ -157,132 +258,271 @@ impl StateExtractor for StarknetState {
-        let address_felt: Felt252 = (*address).into();
-        self.state.state_reader.address_to_class_hash.contains_key(&Address(address_felt))
+        self.state.state.address_to_class_hash.contains_key(address)
     }
 
     fn get_class_hash_at_contract_address(
         &mut self,
         contract_address: &ContractAddress,
     ) -> DevnetResult<ClassHash> {
-        Ok(self.state.get_class_hash_at(&contract_address.into()).map(Felt::new)??)
+        Ok(self
+            .state
+            .get_class_hash_at((*contract_address).try_into()?)
+            .map(|f| Felt::from(f.0))?)
     }
 
-    fn extract_state_diff_from_pending_state(&self) -> DevnetResult<StateDiff> {
-        StateDiff::difference_between_old_and_new_state(
-            self.state.state_reader.as_ref().clone(),
-            self.state.clone(),
-        )
+    fn extract_state_diff_from_pending_state(&mut self) -> DevnetResult<StateDiff> {
+        StateDiff::difference_between_old_and_new_state(self.state.state.clone(), &mut self.state)
     }
 
     fn get_nonce(&self, address: &ContractAddress) -> DevnetResult<Felt> {
         if !self.is_contract_deployed(address) {
             return Err(Error::ContractNotFound);
         }
 
-        self.state.get_nonce_at(&address.into()).map(Felt::from).map_err(Error::from)
+        Ok(self.state.state.nonce_at(address))
     }
 }
 
 #[cfg(test)]
 mod tests {
-    use std::sync::Arc;
-
+    use blockifier::state::state_api::{State, StateReader};
+    use blockifier::test_utils::DictStateReader;
+    use starknet_api::state::StorageKey;
+    use starknet_in_rust::felt::Felt252;
     use starknet_in_rust::services::api::contract_classes::compiled_class::CompiledClass;
-    use starknet_in_rust::state::state_api::{State, StateReader};
-    use starknet_types::cairo_felt::Felt252;
+    use starknet_in_rust::services::api::contract_classes::deprecated_contract_class::ContractClass as StarknetInRustContractClass;
+    use starknet_in_rust::state::in_memory_state_reader::InMemoryStateReader;
     use starknet_types::contract_address::ContractAddress;
-    use starknet_types::contract_class::Cairo0ContractClass;
-    use starknet_types::felt::Felt;
+    use starknet_types::contract_class::{Cairo0ContractClass, ContractClass};
+    use starknet_types::contract_storage_key::ContractStorageKey;
+    use starknet_types::felt::{ClassHash, Felt};
 
-    use super::StarknetState;
-    use crate::error::Error;
+    use super::{DevnetState, StarknetState};
+    use crate::error::{Error, StateError};
     use crate::traits::{StateChanger, StateExtractor};
     use crate::utils::exported_test_utils::dummy_cairo_0_contract_class;
     use crate::utils::test_utils::{
-        dummy_contract_address, dummy_contract_storage_key, dummy_felt,
+        dummy_cairo_1_contract_class, dummy_contract_address, dummy_contract_storage_key,
+        dummy_felt,
     };
 
+    impl starknet_in_rust::state::state_api::StateReader for DevnetState {
```

Comment:
Why is this trait implemented in tests? Just so that we can compare our state implementation with starknet_in_rust?

Reply:
yes